### PR TITLE
feat: add offline mode

### PR DESCRIPTION
**Pull request by Mercy811**
### Summary
Add support for offline. When online, it behaves exactly the same as previous versions. When offline, tracking an event will not schedule a flush not will only be stored in storage. When back to online, the SDK will flush all stored events.

Offline feature is support by default

#### To opt of the default offline logic provided by network checker plugin
```
amplitude.init('API_KEY', {
  offline: OfflineDisabled
});
```

### Checklist

* [ ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:

### CODE REVIEW

File: packages/analytics-browser/src/browser-client.ts [hunk 1]
```diff
@@ -130,6 +133,22 @@ export class AmplitudeBrowser extends AmplitudeCore implements BrowserClient {
     connector.eventBridge.setEventReceiver((event) => {
       void this.track(event.eventType, event.eventProperties);
     });
+
+    // Step 8: Add network listeners for offline mode
```
Comment (justin-fiedler) [inline packages/analytics-browser/src/browser-client.ts:137]: I would prefer if this is done in the Plugin instead. Plugins should  have access to `client.flush()` via `plugin.setup(config, client)`

**Review by justin-fiedler — CHANGES_REQUESTED** [review]

File: packages/analytics-browser/src/browser-client.ts (hunk 1 above)
Comment (Mercy811) [reply packages/analytics-browser/src/browser-client.ts]: I moved all the offline logic to the plugin at https://github.com/amplitude/Amplitude-TypeScript/pull/644/commits/8d14bfc77bb7a7c98fd82309b9cd9a4e990f80bd#diff-0c1d0fcf8156764bdf906a8f23ee6243eff9afeda0799e3551ad83237562e4f0

**Comment by Mercy811** [comment]
It requires `setOffline()` to enable custom offline logic by first opt out the default plugin and then toggling `config.offline` which is unlike Kotlin which has direct access to `amplitude.config`. I think it's an extended feature and we should release this version without supporting custom offline logic first. If there is customer request, then we can plan for it later.

File: packages/analytics-core/test/plugins/destination.test.ts [hunk 2]
```diff
@@ -140,6 +144,40 @@ describe('destination', () => {
       expect(flush).toHaveBeenCalledTimes(2);
     });
 
+    test('should not schedule a flush', async () => {
```
Comment (justin-fiedler) [inline packages/analytics-core/test/plugins/destination.test.ts:147]:
```suggestion
    test('should not schedule a flush if offline', async () => {
```

File: packages/analytics-browser/src/plugins/network-connectivity-checker.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  const setup = async (config: BrowserConfig, amplitude: BrowserClient) => {
+    config.offline = !navigator.onLine;
+
+    addNetworkListener('online', () => {
+      config.offline = false;
+      // Flush immediately will cause ERR_NETWORK_CHANGED
+      setTimeout(() => {
+        amplitude.flush();
+      }, config.flushIntervalMillis);
```
Comment (justin-fiedler) [inline packages/analytics-browser/src/plugins/network-connectivity-checker.ts:43]: I feel this should be more immediate. Can we do next frame instead? `setTimeout(() => {...}, 1)`?

File: packages/analytics-core/src/plugins/destination.ts [hunk 4]
```diff
@@ -117,6 +120,11 @@ export class Destination implements DestinationPlugin {
   }
 
   async flush(useRetry = false) {
+    // Skip flush if offline
+    if (this.config.offline) {
+      return;
```
Comment (justin-fiedler) [inline packages/analytics-core/src/plugins/destination.ts:125]: I feel we should log here to help with debugging. e.g. `config.logger.debug("Skipping flush while offline.")

**Review by justin-fiedler — APPROVED** [review]
Overall looks good. Thanks @Mercy811. Really appreciate the test coverage.

See question about scheduling flush earlier on return to online.

File: packages/analytics-core/src/plugins/destination.ts (hunk 4 above)
Comment (Mercy811) [reply packages/analytics-core/src/plugins/destination.ts:126]: Good point! I added logging messages when goes to online and offline too in https://github.com/amplitude/Amplitude-TypeScript/pull/644/commits/c2c585304a4e6381c648fe471e9603cc5accd735

File: packages/analytics-browser/src/plugins/network-connectivity-checker.ts (hunk 3 above)
Comment (justin-fiedler) [reply packages/analytics-browser/src/plugins/network-connectivity-checker.ts:44]: Ok, I think that's ok. Can you please call this out in docs. Thanks

Comment (Mercy811) [reply packages/analytics-browser/src/plugins/network-connectivity-checker.ts:44]: Definitely!

Comment (Mercy811) [reply packages/analytics-browser/src/plugins/network-connectivity-checker.ts:44]: Docs PR: https://github.com/amplitude/amplitude-dev-center/pull/159

**Comment by Multiply** [comment]
> It requires `setOffline()` to enable custom offline logic by first opting out the default plugin and then toggling `config.offline` which is unlike Kotlin which has direct access to `amplitude.config`. I think adding `setOffline()` is an extended feature and we should release this version without supporting custom offline logic first. If there is customer request, then we can plan for it later.

We just noticed this landing, and we'll most definitely use this functionality to pause sending events, and just let them queue up.
For the most part, we'll just do the same as the default plugin, but also check if we've paused sending events before setting it back to online.

**Comment by Mercy811** [comment]
Hi @Multiply, thanks for choosing Amplitude and using the new feature! I will put it into our roadmap. Could you give me more context on the difference between the default `networkConnectivityCheckerPlugin` and the one you want to implement? We can see if that's what we can make it as default to support more customers with the same needs.

**Comment by Multiply** [comment]
Currently, I'm trying to set offline to disabled, and then set it to true via our own plugin, but it seems to result in the default plugin being loaded anyway. (Our plugin is most likely added first, and setup also called first, which is why the default works fine)

What we're trying to achieve is storing all events in the storage provider, until a user gives consent, instead of simply dropping events.
If the user eventually consents to tracking, we'd like to push the whole queue at that point in time.

Opting out doesn't quite do what we want, but the offline variable would.

Essentially we just want to pause sending for a while.

**Comment by Mercy811** [comment]
Hi @Multiply, 
> What we're trying to achieve is storing all events in the storage provider, until a user gives consent, instead of simply dropping events.
That sounds like another feature request. But glad to hear that we are unblocked now.

**Comment by cgibson-emesent** [comment]
Is there going to be react-native support for offline mode when the network drops out and automatically start sending once the connection returns? I would have expected that if the attempt to send the event fails due to a network error, it's not removed from the _unset array in the local storage. I'm a bit puzzled as to what the purpose of the `offline` config flag is as you can then never send events.

**Comment by Mercy811** [comment]
Hi @Multiply, I created a separate [issue](https://github.com/amplitude/Amplitude-TypeScript/issues/654) to track requests on `setOffline()`. Could you leave a comment with why you need this feature there? Thanks.

**Comment by Mercy811** [comment]
Hi @cgibson-emesent, I also created a separate [issue](https://github.com/amplitude/Amplitude-TypeScript/issues/655) to track requests on supporting offline mode for React Native SDK. Please leave a comment there to help us better track customer requests.

**Pull request merged**